packages/sanity/src/core/form/members/object/fields/ArrayOfObjectsField.tsx

```diff
@@ -58,8 +65,74 @@ export function ArrayOfObjectsField(props: {
     onFieldGroupSelect,
   } = useFormCallbacks()
 
+  const valueRef = useRef(member.field.value)
+  useEffect(() => {
+    valueRef.current = member.field.value
+  }, [member.field.value])
+
+  const handleChange = useCallback(
+    (event: PatchEvent | PatchArg) => {
+      const patches = PatchEvent.from(event).patches
+      // if the patch is an unset patch that targets an item in the array (as opposed to unsetting a field somewhere deeper)
+      const isRemovingLastItem = patches.some(
+        (patch) => patch.type === 'unset' && patch.path.length === 1
+      )
+
+      if (isRemovingLastItem) {
+        // apply the patch to the current value
+        valueRef.current = applyAll(valueRef.current || [], patches)
+
+        // if the result is an empty array
+        if (Array.isArray(valueRef.current) && !valueRef.current.length) {
+          // then unset the array field
+          onChange(PatchEvent.from(unset([member.name])))
+          return
+        }
+      }
+      // otherwise apply the patch
+      onChange(PatchEvent.from(event).prepend(setIfMissing([])).prefixAll(member.name))
+    },
+    [onChange, member.name, valueRef]
+  )
+
+  return (
+    <FormCallbacksProvider
+      onFieldGroupSelect={onFieldGroupSelect}
+      onChange={handleChange}
+      onSetFieldSetCollapsed={onSetFieldSetCollapsed}
+      onSetPathCollapsed={onSetPathCollapsed}
+      onPathOpen={onPathOpen}
+      onPathBlur={onPathBlur}
+      onPathFocus={onPathFocus}
+    >
+      <FieldProvider
+        actions={field.actions}
+        documentId={documentId}
+        documentType={member.field.schemaType.name}
+        focused={focused}
+        path={member.field.path}
+        schemaType={member.field.schemaType}
+      >
+        <Field {...props} onChange={handleChange} />
+      </FieldProvider>
+    </FormCallbacksProvider>
+  )
+}
+
+function Field(props: {
```

Comment:
Might be missing something, but feels like there's very little gain in splitting this up and I would really prefer it if we didn't. It also makes it hard to track the history and changes of this rather critical component (this goes for the other components that had the same split, like `ObjecfField`, `PrimitiveField`, etc

Reply:
The reason for splitting it up is the classic provider-and-need-to-get-react-context in the same component. So it's solved by splitting the children into a separate component, and calling `useFieldActions` there.

Reply:
That seems like an unnecessary indirection to me (again, I might be missing something) - could the action resolving not be done in `ArrayOfObjectsField` in this case?

Reply:
I was able to revert the splitting by doing the field actions resolving in each field wrapper component, instead of in the provider component 👍